End of Game Transition and General Bugfixing

KieranWoodTombola opened the pull request:
Added a score display on timer reaching zero to compare your score to your locally saved highscore (reset on refresh for testing)

Also refined tweens in the background and refined the tweens for selected tokens transitioning on the grid

KieranWoodTombola committed:
Refactored score-display

- seperated grid-related methods within score-display
    into an extended class - GridScoreDisplay - to make the module
    easier to re-use

KieranWoodTombola committed:
Feature: post-game score comparison

- Added a locally stored variable to track highscore

- Added tweens on timer reaching 0 for a comparison between
    current game score and session high score

- Fixed a bug where stage component's events weren't being
    removed. Added class destroy methods for classes
    utilising events to allow for better cleanup

- Still experiencing a bug where, on completing the second session,
    the highscore string presents as currentstring + newstring

KieranWoodTombola committed:
Fix: Fixed Highscore Tween

- Fixed a bug where the Highscore would display the previous
    and new score simultaneously

KieranWoodTombola committed:
Fix: Properly implements ScoreDisplay Callbacks

- ScoreDisplay callbacks were failing to be called because
    they weren't being assigned from the interface in
    the base class constructor

- Fixing this bug has revived the concatenation bug on the
    highScore at the scene level. The current and new highscore are
    displayed simultaneously

KieranWoodTombola committed:
Fix: Post-Game Tweens improved

- Refactored the gsap timeline crammed into the Timer's
timeCompleteCallback into multiple consecutive timelines to better
manage consecutive Tweens

- Made minor changes to modules to allow access to class methods
    at the scene level

- Simplified ScoreDisplay callbacks

KieranWoodTombola committed:
Fix: Refined background wave tweens

- refactored tweens on background so that the transition between
background states result in harsh snapping between tweens

KieranWoodTombola committed:
Fix: Tokens Tween A -> B on curves

- Fixed issue where tokens sharing an X or Y axis didn't tween on a curve

- Reduced the severity of the curve for tokens further apart

KieranWoodTombola committed:
Fix: further reduced severity of tween curve

KieranWoodTombola committed:
Fix: Highscore transition

KieranWoodTombola committed:
Merge branch 'main' of github.com:KieranWoodTombola/Match-Three into gameEnd

KieranWoodTombola committed:
Cleanup: Post-merge cleanup

liamf1986 commented on src/scripts/core/models/score-display.ts:
```diff
@@ -4,63 +4,101 @@ import { Token } from "./token";
+    onScoreChangeComplete: () => void;
+}
+
 export class ScoreDisplay extends Container {
 
-    private _score: number = 0;
-    private _scoreText: PixiText = new PixiText(`${this._score}`);
-    private _textContainer: Container = new Container();
-    private _trackedHeight = -1;
+    public score: number;
+    public scoreAsText: PixiText;
+    public textContainer: Container = new Container();
+    public onScoreChangeComplete: () => void;
```
I couldn't see these variables being accessed outside of the class. Should they be public still?

Just as a general note I try to advocate keeping your class variables private/protected where possible. Providing a getter if something needs to access them, to limit class data being changed from outside accidentally.

KieranWoodTombola replied:
When I make the properties of ScoreDisplay private I get and error on GridScoreDisplay, which extends ScoreDisplay. Is there a better way for me to structure this?

HaydenHowdle replied:
When extending a class and want to access the properties of the class you extend but not make them publicly accessible you want to use the 'protected' access modifier instead of public or private.

```
    protected score: number;
    protected scoreAsText: PixiText;
    protected textContainer: Container = new Container();
    protected onScoreChangeComplete: () => void;
```

Generally you only want to do this if you want extending classes to be able to access and modify these properties but nothing else. Additionally you can make them private and offer a public getter if you need access but don't wish for the extending class to be able to write to the property, there isn't really a right answer here but those are some of your options :)

liamf1986 commented on src/scripts/core/models/background.ts:
```diff
@@ -67,66 +83,54 @@ export class Background extends Container {
+        this.resetSplashTimeline(farWaveHeightMod, midWaveHeightMod ,closeWaveHeightMod, duration);
+    }
 
-        if (this._splash) { this._splash.kill(); }
-        this._splash = gsap.timeline();
```
As we discussed briefly the removal of this looks like your tweens are not being cleared between resets, meaning you're stacking multiple tweens with each reset.

liamf1986 commented on src/scripts/core/services/curve.ts:
```diff
@@ -46,6 +71,7 @@ export class Curve {
         return Math.PI * Math.pow(this.getRadius(), 2);
     }
 
+    //using known degree, get grid position of a point
     public getPointOnCircumference(degrees: number): [number, number] {
```
When commenting functions it can be helpful to use the format `/** comment */`. This will show the comment in intellisense providing that useful information at the point of using the function to help the devs out.

<img width="270" alt="Screenshot 2023-08-25 at 4 50 56 PM" src="https://github.com/KieranWoodTombola/Match-Three/assets/41992522/bfe1a640-e5b9-4f9f-9bd9-c6ce4f38bcae">

<img width="367" alt="Screenshot 2023-08-25 at 4 51 24 PM" src="https://github.com/KieranWoodTombola/Match-Three/assets/41992522/bf87532e-5aa1-4d72-9e10-f4e91edc2faa">

AndrewMollett replied:
It also allows you to spread the comment across multiple lines, as well as insert additional things such as @param which let's you describe what a parameter is for, letting you customise what intellisense shows when invoking said function. Similarly you can give a type to the param by using {} notation. Not always handy as the name of a parameter can suffice, but it can allow for a little extra description where needed.
<img width="489" alt="image" src="https://github.com/KieranWoodTombola/Match-Three/assets/13646039/165bf20a-61c7-414d-b0c7-ad5456831410">
<img width="327" alt="image" src="https://github.com/KieranWoodTombola/Match-Three/assets/13646039/6184d852-08aa-44d0-9d55-2b5d2bd044b2">

liamf1986 commented on src/scripts/scenes/game-scene.ts:
```diff
@@ -47,52 +47,167 @@ export class GameScene extends Scene {
+            .to(highScoreDisplay, {
+                x: _viewWidth * 0.5 - highScoreDisplay.width * 0.5,
+            }, 1)
+            .to(highScoreDisplay, {
+            }, 1);
```
Looks like this last one isnt doing anything?

liamf1986 commented on src/scripts/core/models/background.ts:
```diff
@@ -67,66 +83,54 @@ export class Background extends Container {
-            yoyo: true
-        }, 0);
+    private resetSplashTimeline(farWaveHeightMod: number, midWaveHeightMod: number, closeWaveHeightMod: number, duration: number): void {
+        this._splash.then( () => {
+            this._splash.yoyo(false);
+        })
+        .then(() => {
+            this.addWavesToTimeline(farWaveHeightMod, midWaveHeightMod, closeWaveHeightMod, duration);
+        });
```
As mentioned I hadn't realised you could use `.then` from a Timeline so this was new to me. Had a little look though and it is just a Promise implementation of the onComplete approach. So the 2 should be interchangable.

With the chaining here you're not getting any thing from the second `.then` and can instead put it all in one like so:
```ts
this._splash.then(() => {
    this.splash.yoyo(false);
   this.addWavesToTimeline(etc...);
});
```

The ability to chain `.then` together is for when you have other asynchronous code you want to run inside the `.then` before the next `.then` should happen i.e.

```ts
this._splash.then(() => {
    return requestSomeDataFromServer();
}).then((response) => {
    // this would happen after receiving the response from the server
    console.log(response);
});
```

KieranWoodTombola replied:
Aha! That makes more sense now
I'll make the change

KieranWoodTombola replied:
Actually, on further testing, I think I'll drop the changes to wave height.
There's not really enough room on the stage for the transitions. Even when they work as intended, big waves don't look very good.

KieranWoodTombola committed:
PR: Responding to changes

KieranWoodTombola committed:
PR: Responding to changes

- Updated ScoreDisplay property modifiers

- Updated description of getPointOnCircumference method
in the Curve class
